odoo/addons/base/models/ir_default.py

```diff
@@ -131,16 +131,20 @@ def get_model_defaults(self, model_name, condition=False):
                     JOIN res_users u ON u.id=%s
                     WHERE f.model=%s
                         AND (d.user_id IS NULL OR d.user_id=u.id)
-                        AND (d.company_id IS NULL OR d.company_id=u.company_id)
-                        AND {}
+                        AND (d.company_id IS NULL OR d.company_id={id1})
+                        AND {id2}
                     ORDER BY d.user_id, d.company_id, d.id
                 """
+        default_company = self._context.get('default_company_id')
```

Comment:
By doing this you're making the defaults dependent on the context, so the ormcache invalidation would need to be updated.